code/contracts/ONEWallet2.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+contract ONEWallet2 {
+
+    bytes32 root; // Note: @ivan brought up a good point in reducing this to 16-bytes so hash of two consecutive nodes can be done in a single word (to save gas and reduce blockchain clutter). Let's not worry about that for now and re-evalaute this later.
+    uint8 height; // including the root. e.g. for a tree with 4 leaves, the height is 3.
+    uint8 interval; // otp interval in seconds, default is 30
+    uint32 t0; // starting time block (effectiveTime (in ms) / interval)
+    uint32 lifespan;  // in number of blocks (e.g. 1 block per [interval] seconds)
+    
+    address payable lastResortAddress; // where money will be sent during a recovery process (or when the wallet is beyond its lifespan)
+    bytes32 recoveryHash; // hash of the secret stored at the client, which can be used to recover the walletif the user loses the authenticator
+    uint8 constant MAX_REVEAL_DELAY = 2; // in OTP intervals (i.e., 2 * interval seconds)    
+    
+    // Management of daily spendings
+    uint256 dailyLimit; // uint128 is sufficient, but uint256 is more efficient since EVM works with 32-byte words.
+    mapping(uint => uint) dailySpendings; // day => amount ; can be cleaned up once in a while to save storage
+    uint cleanupCheckpoint; // the index of the day since epoch, when the last cleanup was made
+    uint16 constant CLEAN_UP_DAILY_SPENDING_BATCH = 1024; // this is to avoid out-of-gas exceptions
+    uint8 AUTO_CLEANUP_DAYS = 100; // do the auto cleanup of daily spending data after every 100 days
+    
+    uint lastExecutedIdx; // stores the index of OTP that was executed last (this is to avoid reply attacks within the same OTP interval by not allowing commits <= than this number)
+
+    enum OperType { TRANSFER, RECOVERY, SET_LAST_RESORT_ADDR }
+
+    struct Commit {
+        bytes32 hash;          // h(eOTP || operType || params)  
+        uint timestamp;      // OTP must be valid in this time; not in the time of reveal        
+        OperType operType;
+        // below are operation-specific parameters, based on operType
+        address payable addr; 
+        uint amount;        
+    }    
+    mapping(uint32 => Commit[]) public commits; // idx of OTP valid during a commitment => commitment data
```

Comment:
I noticed only revealed commits are deleted (line 103). The commits would grow unbounded if an attacker spams it.

Reply:
It clutters the blockchain. But another potential danger is if we are to use an upgrade pattern in this contract. The attacker who is allowed to make unlimited commits may find a way to override key storage locations through spamming commits, though the exact formulation of the attack is unknown at this time due to the assumed reliability of pseudo-randomness of SHA3.

Reply:
I'd say we can safely assume that sha3 is collision-resistant and thus provides enough pseudo-randomness.

Reply:
Proper clean-up method was already added.